src/Error/ErrorTrap.php

```diff
@@ -0,0 +1,254 @@
+    {
+        $this->setConfig($options);
+        $this->setErrorRenderer($this->chooseErrorRenderer());
+        $this->setLogger($this->getConfig('logger'));
+        $this->setLevel($this->getConfig('errorLevel'));
+    }
+
+    /**
+     * Choose an error renderer based on config or the SAPI
+     *
+     * @return class-string<\Cake\Error\ErrorRendererInterface>
+     */
+    protected function chooseErrorRenderer(): string
+    {
+        $config = $this->getConfig('errorRenderer');
+        if ($config !== null) {
+            return $config;
+        }
+
+        /** @var class-string<\Cake\Error\ErrorRendererInterface> */
+        return PHP_SAPI === 'cli' ? ConsoleRenderer::class : HtmlRenderer::class;
+    }
+
+    /**
+     * Attach this ErrorTrap to PHP's default error handler.
+     *
+     * This will replace the existing error handler, and the
+     * previous error handler will be discarded.
+     *
+     * @return void
+     */
+    public function register(): void
+    {
+        $this->registered = true;
+
+        $level = $this->_config['errorLevel'] ?? -1;
+        error_reporting($level);
+        set_error_handler([$this, 'handleError'], $level);
+    }
+
+    /**
+     * Handle an error from PHP set_error_handler
+     *
+     * Will use the configured renderer to generate output
+     * and output it.
+     *
+     * @param int $code Code of error
+     * @param string $description Error description
+     * @param string|null $file File on which error occurred
+     * @param int|null $line Line that triggered the error
+     * @return bool True if error was handled
+     */
+    public function handleError(
+        int $code,
+        string $description,
+        ?string $file = null,
+        ?int $line = null
+    ): bool {
+        if (!(error_reporting() & $code)) {
+            return false;
+        }
+        $trace = Debugger::trace(['start' => 1, 'format' => 'points']);
+        $error = new PhpError($code, $description, $file, $line, $trace);
+
+        $renderer = $this->renderer();
+        $logger = $this->logger();
+
+        try {
+            // Log first incase rendering or callbacks fail.
+            $logger->logMessage($error->getLabel(), $error->getMessage());
+
+            foreach ($this->callbacks as $callback) {
+                $callback($error);
+            }
+            $renderer->write($renderer->render($error));
+        } catch (Exception $e) {
+            $logger->logMessage('error', 'Could not render error. Got: ' . $e->getMessage());
+
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Get an instance of the renderer.
+     *
+     * @return \Cake\Error\ErrorRendererInterface
+     */
+    public function renderer(): ErrorRendererInterface
+    {
+        $class = $this->getConfig('errorRenderer');
+
+        /** @var \Cake\Error\ErrorRendererInterface $instance */
+        $instance = new $class($this->_config);
+
+        return $instance;
```

Comment:
Feels a little odd to have this always instantiate. Would you ever need to get the renderer in multiple calls?

Reply:
I vote for simplest design. Just trying to discuss expected convention.

Reply:
If the renderers or logger constructor setup logic then I might be concerned with the overhead. Retaining instances means we need attributes, and methods to unset the instances if/when class names change. I think not retaining instances is the simpler design.